### PR TITLE
DYN-7122: Node autocomplete clusters - internal testing tool

### PR DESCRIPTION
**Pull request by reddyashish**
### Purpose

Based on the spike: https://jira.autodesk.com/browse/DYN-7122

Using the node autocomplete view extension to interact with the results from Node autocomplete cluster service. 
We will be introducing a new preference setting for the node autocomplete cluster and when user turns that on, we enable the Node autocomplete cluster functionality. Double clicking on the port will show the results in the view extension. 
More user action items will be implemented in next tasks.

NDA cluster

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB

### Release Notes

(FILL ME IN) Brief description of the fix / enhancement. Use N/A to indicate that the changes in this pull request do not apply to Release Notes. **Mandatory section**

### Reviewers

(FILL ME IN) Reviewer 1 (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: src/DynamoCore/Search/NodeAutocompleteSearch.cs [hunk 1]
```diff
@@ -35,49 +37,52 @@ internal MLNodeAutoCompletionRequest(string dynamoVersion, int numberOfResults)
...
-        internal ContextRequest Context { get; set; }
+        internal ContextItem Context { get; set; }
     }
 
+    /// <summary>
+    /// Data representing node properties for ML autocomplete.
+    /// </summary>
     [DataContract]
-    internal class NodeRequest
+    public class NodeItem
```
Comment (QilongTang) [inline src/DynamoCore/Search/NodeAutocompleteSearch.cs:47]: If it's internal visible, do we still need to make it public?

File: src/DynamoCore/Search/NodeAutocompleteSearch.cs [hunk 2]
```diff
@@ -188,21 +213,77 @@ internal MLNodeAutoCompletionResponse()
         internal IEnumerable<ResultItem> Results { get; set; }
     }
 
+    [DataContract]
+    internal class MLNodeClusterAutoCompletionResponse
```
Comment (QilongTang) [inline src/DynamoCore/Search/NodeAutocompleteSearch.cs:217]: Some more comments?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 3]
```diff
@@ -219,6 +219,17 @@ internal bool IsDNADataIngestionPipelineinBeta
...
         }
 
+        /// <summary>
+        /// Controls if the cluster node autocomplete placement feature is enabled from feature flag
+        /// </summary>
+        internal bool IsNodeAutocompleteClusterEnabled
+        {
+            get
+            {
+                return DynamoModel.FeatureFlags?.CheckFeatureFlag("IsNodeAutocompleteClusterEnabled", false) ?? true;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:229]: 👍🏻

File: src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs [hunk 4]
```diff
@@ -416,6 +420,28 @@ private void AutoComplete(object parameter)
...
+            {
+                wsViewModel.DynamoViewModel.Model.CurrentWorkspace.Undo();
+            }
+
+            // Bail out from connect state
+            wsViewModel.CancelActiveState();
+
+            if (wsViewModel.DynamoViewModel.IsNodeAutocompleteClusterEnabled)
+            {
+               MLNodeClusterAutoCompletionResponse results = wsViewModel.NodeAutoCompleteSearchViewModel.GetMLNodeClusterAutocompleteResults();
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs:440]: Ok, so the results are still stored in `NodeAutoCompleteSearchViewModel` side and access by view extension. The view extension itself does not handle any communication layer to the service. Would you cover that in the design docs you are working on right now?

File: src/DynamoCore/Search/NodeAutocompleteSearch.cs (hunk 2 above)
Comment (reddyashish) [reply src/DynamoCore/Search/NodeAutocompleteSearch.cs:240]: Added comments.

File: src/DynamoCore/Search/NodeAutocompleteSearch.cs (hunk 1 above)
Comment (reddyashish) [reply src/DynamoCore/Search/NodeAutocompleteSearch.cs]: Actually this object was being used in the NodeAutocompleteCluster(this had to be public) which is exposed in the view extension. But for now I can make it internal as we are not previewing nodes/connectors. Can make it public later if needed.

File: src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs (hunk 4 above)
Comment (reddyashish) [reply src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs:440]: So the results are actually stored in NodeAutocompleteCluster which is in NodeAutoCompleteView.xaml.cs. The API to call the service is in the search view model, which i plan to move to the NodeAutoCompleteViewModel.cs(introduced in this PR)

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 3 above)
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:229]: @reddyashish I found we already have one feature flag for it - `IsDNAClusterPlacementEnabled`, default value is on

Comment (reddyashish) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: updated it. Thanks Aaron

**Comment by reddyashish** [comment]
@QilongTang Verified that this is only enabled in the dev builds.

File: src/NodeAutoCompleteViewExtension/NodeAutoCompleteViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+        internal DynamoViewModel dynamoViewModel { get; set; }
+
+        //constructor
+        internal NodeAutoCompleteViewModel(UIElement dynamoView, DynamoViewModel dvm)
+        {
+            DynamoView = dynamoView;
+            dynamoViewModel = dvm;
+        }
+
+        /*internal MLNodeAutoCompletionRequest GenerateRequestForMLAutocomplete()
```
Comment (QilongTang) [inline src/NodeAutoCompleteViewExtension/NodeAutoCompleteViewModel.cs:39]: @reddyashish What are these code commented out for?

Comment (reddyashish) [reply src/NodeAutoCompleteViewExtension/NodeAutoCompleteViewModel.cs]: Forgot to remove it. It is duplicated code.

**Review by QilongTang — APPROVED** [review]

**Comment by reddyashish** [comment]
Passed: https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/16910/

**Pull request merged**